testing/asserts: `assertAlmostEquals()`

<!-- NOTE: for larger scale feature requests like additions to the standard library, consider opening a new discussion under the "Proposals" category over here: https://github.com/denoland/deno_std/discussions/new?category=proposals -->

**Is your feature request related to a problem? Please describe.**

I'm writing unit tests for a lot of math heavy functions, so floating point errors are all over the place. For instance:
```js
import { assertEquals } from "https://deno.land/std@0.125.0/testing/asserts.ts";

const result = sum(0.1, 0.2);

assertEquals(result, 0.3);
```
results in
```
Uncaught AssertionError: Values are not equal:

    [Diff] Actual / Expected

-   0.30000000000000004
+   0.3
```
**Describe the solution you'd like**

A method for comparing numbers with a tolerance:

```js
assertAlmostEquals(result, 0.3, 0.0001 /* tolerance */);
```

alternative names:
- `assertCloseTo()` (from [chai](https://www.chaijs.com/api/bdd/#closetoexpected-delta-msg))
- `assertAlmostEqual()` or `assertAlmostEquals()` (from [Python](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertAlmostEqual))
- `assertRoughlyEquals()`

**Describe alternatives you've considered**

I'll use a helper function included with the tests, I'm working with vector math so I'm going to need a helper function for comparing vectors anyway, but I figured std could benefit from something like this as well.

--- a/testing/README.md
+++ b/testing/README.md
@@ -18,6 +18,9 @@ pretty-printed diff of failing assertion.
   and `expected` are equal.
 - `assertStrictEquals()` - Compares `actual` and `expected` strictly, therefore
   for non-primitives the values must reference the same instance.
+- `assertAlmostEquals()` - Make an assertion that `actual` is almost equal to
+  `expected`, according to a given `epsilon` _(defaults
+  to `1e-7`)_
 - `assertStringIncludes()` - Make an assertion that `actual` includes
   `expected`.
 - `assertMatch()` - Make an assertion that `actual` match RegExp `expected`.

--- a/testing/README.md
+++ b/testing/README.md
@@ -19,8 +19,7 @@ pretty-printed diff of failing assertion.
 - `assertStrictEquals()` - Compares `actual` and `expected` strictly, therefore
   for non-primitives the values must reference the same instance.
 - `assertAlmostEquals()` - Make an assertion that `actual` is almost equal to
-  `expected`, according to a given `epsilon` _(defaults
-  to `1e-7`)_
+  `expected`, according to a given `epsilon` _(defaults to `1e-7`)_
 - `assertStringIncludes()` - Make an assertion that `actual` includes
   `expected`.
 - `assertMatch()` - Make an assertion that `actual` match RegExp `expected`.

--- a/testing/asserts.ts
+++ b/testing/asserts.ts
@@ -418,7 +418,7 @@ export function assertNotStrictEquals(
  * If the values are not almost equal then throw.
  *
  * ```ts
- * import { assertAlmostEquals } from "./asserts.ts";
+ * import { assertAlmostEquals, assertThrows } from "./asserts.ts";
  *
  * assertAlmostEquals(0.1, 0.2);
  *

--- a/testing/asserts.ts
+++ b/testing/asserts.ts
@@ -420,7 +420,11 @@ export function assertNotStrictEquals(
  * ```ts
  * import { assertAlmostEquals } from "./asserts.ts";
  *
- * assertAlmostEquals(0.1, 0.2)
+ * assertAlmostEquals(0.1, 0.2);
+ * 
+ * // Using a custom epsilon value
+ * assertAlmostEquals(0.1 + 0.2, 0.3, 1e-16);
+ * assertThrows(() => assertAlmostEquals(0.1 + 0.2, 0.3, 1e-17));
  * ```
  */
 export function assertAlmostEquals(

--- a/testing/asserts.ts
+++ b/testing/asserts.ts
@@ -432,16 +432,16 @@ export function assertAlmostEquals(
   if (actual === expected) {
     return;
   }
-  if (
-    (Number.isNaN(actual)) || (Number.isNaN(expected)) ||
-    (Math.abs(expected - actual) > epsilon)
-  ) {
-    const f = (n: number) =>
-      n.toFixed(-Math.log10(epsilon)).replace(/\.?0+$/, "");
-    throw new AssertionError(
-      msg ?? `actual: "${f(actual)}" expected to be close to "${f(expected)}"`,
-    );
+  const delta = Math.abs(expected - actual);
+  if (delta <= epsilon) {
+    return;
   }
+  const f = (n: number) => Number.isInteger(n) ? n : n.toExponential();
+  throw new AssertionError(
+    msg ??
+      `actual: "${f(actual)}" expected to be close to "${f(expected)}": \
+delta "${f(delta)}" is greater than "${f(epsilon)}"`,
+  );
 }
 
 /**

--- a/testing/asserts.ts
+++ b/testing/asserts.ts
@@ -412,6 +412,36 @@ export function assertNotStrictEquals(
 }
 
 /**
+ * Make an assertion that `actual` and `expected` are almost equal numbers through
+ * a given epsilon. It can be used to take into account IEEE-754 double-precision
+ * floating-point representation limitations.
+ * If the values are not almost equal then throw.
+ *
+ * ```ts
+ * import { assertAlmostEquals } from "./asserts.ts";
+ *
+ * assertAlmostEquals(0.1, 0.2)
+ * ```
+ */
+export function assertAlmostEquals(
+  actual: number,
+  expected: number,
+  epsilon = 1e-7,
+  msg?: string,
+) {
+  if (actual === expected) {
+    return;
+  }
+  if (Math.abs(expected - actual) > epsilon) {
+    const f = (n: number) =>
+      n.toFixed(-Math.log10(epsilon)).replace(/\.?0+$/, "");
+    throw new AssertionError(
+      msg ?? `actual: "${f(actual)}" expected to be close to "${f(expected)}"`,
+    );
+  }
+}
+
+/**
  * Make an assertion that actual is not null or undefined.
  * If not then throw.
  */

--- a/testing/asserts.ts
+++ b/testing/asserts.ts
@@ -413,7 +413,7 @@ export function assertNotStrictEquals(
 
 /**
  * Make an assertion that `actual` and `expected` are almost equal numbers through
- * a given epsilon. It can be used to take into account IEEE-754 double-precision
+ * a given tolerance. It can be used to take into account IEEE-754 double-precision
  * floating-point representation limitations.
  * If the values are not almost equal then throw.
  *
@@ -422,7 +422,7 @@ export function assertNotStrictEquals(
  *
  * assertAlmostEquals(0.1, 0.2);
  *
- * // Using a custom epsilon value
+ * // Using a custom tolerance value
  * assertAlmostEquals(0.1 + 0.2, 0.3, 1e-16);
  * assertThrows(() => assertAlmostEquals(0.1 + 0.2, 0.3, 1e-17));
  * ```
@@ -430,21 +430,21 @@ export function assertNotStrictEquals(
 export function assertAlmostEquals(
   actual: number,
   expected: number,
-  epsilon = 1e-7,
+  tolerance = 1e-7,
   msg?: string,
 ) {
   if (actual === expected) {
     return;
   }
   const delta = Math.abs(expected - actual);
-  if (delta <= epsilon) {
+  if (delta <= tolerance) {
     return;
   }
   const f = (n: number) => Number.isInteger(n) ? n : n.toExponential();
   throw new AssertionError(
     msg ??
       `actual: "${f(actual)}" expected to be close to "${f(expected)}": \
-delta "${f(delta)}" is greater than "${f(epsilon)}"`,
+delta "${f(delta)}" is greater than "${f(tolerance)}"`,
   );
 }
 

--- a/testing/asserts.ts
+++ b/testing/asserts.ts
@@ -421,7 +421,7 @@ export function assertNotStrictEquals(
  * import { assertAlmostEquals } from "./asserts.ts";
  *
  * assertAlmostEquals(0.1, 0.2);
- * 
+ *
  * // Using a custom epsilon value
  * assertAlmostEquals(0.1 + 0.2, 0.3, 1e-16);
  * assertThrows(() => assertAlmostEquals(0.1 + 0.2, 0.3, 1e-17));

--- a/testing/asserts.ts
+++ b/testing/asserts.ts
@@ -432,7 +432,10 @@ export function assertAlmostEquals(
   if (actual === expected) {
     return;
   }
-  if (Math.abs(expected - actual) > epsilon) {
+  if (
+    (Number.isNaN(actual)) || (Number.isNaN(expected)) ||
+    (Math.abs(expected - actual) > epsilon)
+  ) {
     const f = (n: number) =>
       n.toFixed(-Math.log10(epsilon)).replace(/\.?0+$/, "");
     throw new AssertionError(

--- a/testing/asserts_test.ts
+++ b/testing/asserts_test.ts
@@ -2,6 +2,7 @@
 import {
   _format,
   assert,
+  assertAlmostEquals,
   assertArrayIncludes,
   assertEquals,
   assertExists,
@@ -912,6 +913,23 @@ Deno.test({
   fn(): void {
     assertThrows(() => assertNotStrictEquals(1, 1), AssertionError);
   },
+});
+
+Deno.test("assert almost equals number", () => {
+  //Default precision
+  assertAlmostEquals(-0, +0);
+  assertAlmostEquals(Math.PI, Math.PI);
+  assertAlmostEquals(0.1 + 0.2, 0.3);
+  assertThrows(() => assertAlmostEquals(1, 2));
+  assertThrows(() => assertAlmostEquals(1, 1.1));
+
+  //Higher precision
+  assertAlmostEquals(0.1 + 0.2, 0.3, 1e-16);
+  assertThrows(
+    () => assertAlmostEquals(0.1 + 0.2, 0.3, 1e-17),
+    AssertionError,
+    '"0.30000000000000004" expected to be close to "0.29999999999999999"',
+  );
 });
 
 Deno.test({

--- a/testing/asserts_test.ts
+++ b/testing/asserts_test.ts
@@ -930,6 +930,29 @@ Deno.test("assert almost equals number", () => {
     AssertionError,
     '"0.30000000000000004" expected to be close to "0.29999999999999999"',
   );
+
+  //Special cases
+  assertAlmostEquals(Infinity, Infinity);
+  assertThrows(
+    () => assertAlmostEquals(0, Infinity),
+    AssertionError,
+    '"0" expected to be close to "Infinity"',
+  );
+  assertThrows(
+    () => assertAlmostEquals(-Infinity, +Infinity),
+    AssertionError,
+    '"-Infinity" expected to be close to "Infinity"',
+  );
+  assertThrows(
+    () => assertAlmostEquals(Infinity, NaN),
+    AssertionError,
+    '"Infinity" expected to be close to "NaN"',
+  );
+  assertThrows(
+    () => assertAlmostEquals(NaN, NaN),
+    AssertionError,
+    '"NaN" expected to be close to "NaN"',
+  );
 });
 
 Deno.test({

--- a/testing/asserts_test.ts
+++ b/testing/asserts_test.ts
@@ -928,7 +928,9 @@ Deno.test("assert almost equals number", () => {
   assertThrows(
     () => assertAlmostEquals(0.1 + 0.2, 0.3, 1e-17),
     AssertionError,
-    '"0.30000000000000004" expected to be close to "0.29999999999999999"',
+    `"${(0.1 + 0.2).toExponential()}" expected to be close to "${
+      (0.3).toExponential()
+    }"`,
   );
 
   //Special cases

